File: deploy_kr260/tests/test_stream_adder.py
#!/usr/bin/env python3
"""
test_stream_adder.py - Test completo del IP stream_adder en KR260

Requiere:
  - xmutil loadapp stream_adder (ya ejecutado)
  - sudo (para acceso a /dev/mem)

Este script:
  1. Abre /dev/mem y mapea el bloque AXI-Lite del stream_adder
  2. Escribe distintos valores en el registro add_value (reg0)
  3. Los lee y verifica que se escribieron correctamente
  4. Prueba casos borde (0, 0xFFFFFFFF, valores intermedios)

Uso:
  sudo python3 test_stream_adder.py
"""

import mmap
import os
import struct
import sys

# Direcciones del AXI-Lite del stream_adder
STREAM_ADDER_BASE = 0xA0010000
STREAM_ADDER_SIZE = 0x1000   # 4 KB

# Registros (offsets desde base)
REG_ADD_VALUE = 0x00    # reg0 - valor a sumar al stream


def test_register_rw(mem, name, offset, test_values):
    """Prueba escribir y leer un registro."""
    print(f"\n=== Test registro {name} (offset 0x{offset:02X}) ===")
    passed = 0
    failed = 0

    for val in test_values:
        # Escribir
        mem.seek(offset)
        mem.write(struct.pack("<I", val))

        # Leer
        mem.seek(offset)
        read_back = struct.unpack("<I", mem.read(4))[0]

        status = "OK" if read_back == val else "FAIL"
        if read_back == val:
            passed += 1
        else:
            failed += 1

        print(f"  write 0x{val:08X} -> read 0x{read_back:08X}  [{status}]")

    return passed, failed


def main():
    if os.geteuid() != 0:
        print("ERROR: este script requiere sudo (acceso a /dev/mem)")
        sys.exit(1)

    try:
        with open("/dev/mem", "r+b") as f:
            mem = mmap.mmap(
                f.fileno(),
                STREAM_ADDER_SIZE,
                mmap.MAP_SHARED,
                mmap.PROT_READ | mmap.PROT_WRITE,
                offset=STREAM_ADDER_BASE,
            )
    except Exception as e:
        print(f"ERROR abriendo /dev/mem: {e}")
        print("  -> Comprobar que xmutil loadapp stream_adder ejecuto OK")
        print("  -> cat /sys/class/fpga_manager/fpga0/state debe decir 'operating'")
        sys.exit(1)

    print(f"AXI-Lite mapeado en 0x{STREAM_ADDER_BASE:08X} - 0x{STREAM_ADDER_BASE+STREAM_ADDER_SIZE:08X}")

    # Valores a probar
    test_vals = [
        0x00000000,
        0x00000001,
        0x0000000F,
        0x000000FF,
        0x0000FFFF,
        0x12345678,
        0xDEADBEEF,
        0xFFFFFFFF,
    ]

    passed, failed = test_register_rw(mem, "add_value", REG_ADD_VALUE, test_vals)

    print(f"\n{'='*50}")
    print(f"RESULTADO: {passed} OK, {failed} FAIL de {len(test_vals)} pruebas")
    print(f"{'='*50}")

    mem.close()

    sys.exit(0 if failed == 0 else 1)


if __name__ == "__main__":
    main()

File: deploy_kr260/tests/test_stream_adder.c
/*
 * test_stream_adder.c - Test AXI-Lite del stream_adder en KR260
 *
 * Compilar en la placa:
 *   gcc -O2 -Wall -o test_stream_adder test_stream_adder.c
 *
 * Ejecutar:
 *   sudo ./test_stream_adder
 *
 * Requiere: xmutil loadapp stream_adder ya ejecutado
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/mman.h>

#define STREAM_ADDER_BASE  0xA0010000UL
#define STREAM_ADDER_SIZE  0x1000UL
#define REG_ADD_VALUE      0x00

static volatile uint32_t *mem;

static void reg_write(uint32_t off, uint32_t val) {
    mem[off / 4] = val;
    __sync_synchronize();
}

static uint32_t reg_read(uint32_t off) {
    __sync_synchronize();
    return mem[off / 4];
}

int main(void) {
    int fd = open("/dev/mem", O_RDWR | O_SYNC);
    if (fd < 0) {
        perror("open /dev/mem");
        fprintf(stderr, "  -> necesitas sudo\n");
        return 1;
    }

    void *map = mmap(NULL, STREAM_ADDER_SIZE, PROT_READ | PROT_WRITE,
                     MAP_SHARED, fd, STREAM_ADDER_BASE);
    if (map == MAP_FAILED) {
        perror("mmap");
        close(fd);
        return 1;
    }
    mem = (volatile uint32_t *)map;

    printf("Mapeado AXI-Lite en 0x%08lX\n", STREAM_ADDER_BASE);

    static const uint32_t vals[] = {
        0x00000000, 0x00000001, 0x000000FF, 0x0000FFFF,
        0x12345678, 0xDEADBEEF, 0xFFFFFFFF
    };

    int pass = 0, fail = 0;
    for (size_t i = 0; i < sizeof(vals) / sizeof(vals[0]); i++) {
        reg_write(REG_ADD_VALUE, vals[i]);
        uint32_t r = reg_read(REG_ADD_VALUE);
        if (r == vals[i]) {
            printf("  [OK]   write 0x%08X read 0x%08X\n", vals[i], r);
            pass++;
        } else {
            printf("  [FAIL] write 0x%08X read 0x%08X\n", vals[i], r);
            fail++;
        }
    }

    printf("\nRESULTADO: %d OK, %d FAIL\n", pass, fail);

    munmap(map, STREAM_ADDER_SIZE);
    close(fd);

    return fail == 0 ? 0 : 1;
}
